stage0/src/dice_attestation.rs

```diff
@@ -0,0 +1,129 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+extern crate alloc;
+
+use alloc::{string::String, vec::Vec};
+use coset::{cbor::value::Value, cwt, iana, CborSerializable, CoseError};
+use hkdf::Hkdf;
+use p384::ecdsa::{signature::Signer, Signature, SigningKey, VerifyingKey};
+use rand_core::OsRng;
+use sha2::Sha256;
+
+
+pub const ID_SALT: &str = "DICE_ID_SALT";
+pub const SIGNATURE_LENGTH: usize = 20;
+
+/// Attestation related functions
+///
+/// Returns Signed Stage1 key and measurments
+struct Stage0Signer {
+    signing_key: SigningKey,
+}
+
+// Signer implementation.
+impl Stage0Signer {
+    fn sign(&self, data: &[u8]) -> Vec<u8> {
+        let signed_stage1_ca_verifying_key: Signature = self.signing_key.sign(data);
+        Vec::from(signed_stage1_ca_verifying_key.to_bytes().as_slice())
+    }
+}
+
+fn derive_id(ikm: &[u8], info: Option<&[u8]>) -> [u8; SIGNATURE_LENGTH] {
+    let hkdf = Hkdf::<Sha256>::new(Some(ID_SALT.as_bytes()), ikm);
+    let mut okm: [u8; SIGNATURE_LENGTH] = [0u8; SIGNATURE_LENGTH];
+    hkdf.expand(info.unwrap_or(&[]), &mut okm)
+        .expect("20 is a valid length for Sha256 to output");
+
+    okm
+}
+
+fn generate_ecdsa_keys(info: &str) -> (SigningKey, [u8; SIGNATURE_LENGTH]) {
```

Comment:
I would keep this as two different functions. Combining key generation and ID generation makes the function unintuitive. Or, at least it should have detailed doc comments describing the inputs, outputs  and intended usage.

Reply:
+1